src/v/cloud_storage/partition_manifest.cc

```diff
@@ -328,47 +455,80 @@ partition_manifest::difference(const partition_manifest& remote_set) const {
       end(),
       remote_set.begin(),
       remote_set.end(),
-      result.get_insert_iterator());
+      conditional_inserter(
+        [&remote_set](const auto& kv) {
+            // Insert segment only if it's not present in the
+            // replaced set. Note that the replaced collection is
+            // a multimap, so we need an extra check.
+            auto it = remote_set._replaced.find(kv.first);
+            return it == remote_set._replaced.end()
+                   || it->second.committed_offset != kv.second.committed_offset;
+        },
+        result._segments,
+        result._segments.begin()));
+    std::set_difference(
+      _replaced.begin(),
+      _replaced.end(),
+      remote_set._replaced.begin(),
+      remote_set._replaced.end(),
+      conditional_inserter(
+        [&remote_set](const auto& kv) {
+            // Insert 'replaced' only if it's not present in
+            // the remote_set manifest (represents new data)
+            return remote_set.find(kv.first.base_offset) == remote_set.end();
```

Comment:
nit: would it be possible to use `remote_set.contains(kv.first.base_offset)` here for conciseness?